File: Manager/UIComposer/src/main/webapp/js/controller/screenViewController.js
/*
 * OpenRemote, the Home of the Digital Home. Copyright 2008, OpenRemote Inc.
 *
 * See the contributors.txt file in the distribution for a full listing of individual contributors.
 *
 * This is free software; you can redistribute it and/or modify it under the terms of the GNU General Public License as
 * published by the Free Software Foundation; either version 3.0 of the License, or (at your option) any later version.
 *
 * This software is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without even the implied
 * warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
 *
 * You should have received a copy of the GNU General Public License along with this software; if not, write to the Free
 * Software Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA, or see the FSF site:
 * http://www.fsf.org.
 */

ScreenViewController = function() {
    function ScreenViewController() {
		
    }

    /**
     * Show create Screen dialog.
     */
    function showCreateScreenDialog() {
        $("#create_screen_dialog").showModalForm("Create Screen", {
            'Create': confirmCreateScreen
        });

        $("#create_screen_dialog").enterKeyPressed(confirmCreateScreen);
    }

    /**
     * Invoked when user confirm create Screen.
     */
    function confirmCreateScreen() {
        var name = $("#screen_name_input");
        var row = $("#screen_row_input");
        var col = $("#screen_col_input");
        var valid = true;
        if ($.empty(name.val())) {
            valid = false;
            $("#create_screen_dialog").updateTips(name, "Name is required");
            return;
        }
        if ($.empty(row.val())) {
            valid = false;
            $("#create_screen_dialog").updateTips(row, "Row is required");
            return;
        }
        if (!row.val().toString().isNumber()) {
            valid = false;

            $("#create_screen_dialog").updateTips(row, "Row must be a number");
            return;
        }else if(row.val()>6){
            valid = false;

            $("#create_screen_dialog").updateTips(row, "Row must be less than or equal 6");
            return;
        }

        if ($.empty(col.val())) {
            valid = false;
            $("#create_screen_dialog").updateTips(col, "Column is required");
            return;
        }
        if (!col.val().toString().isNumber()) {
            valid = false;
            $("#create_screen_dialog").updateTips(col, "Column must be a number");
            return;
        }else if(col.val()>4){
            valid = false;

            $("#create_screen_dialog").updateTips(col, "Column must be less than or equal 4");
            return;
        }
        if (valid) {
            var screen = new Screen();
            screen.id = global.BUTTONID++;
            screen.name = name.val();
			screen.row = row.val();
			screen.col = col.val();
            ScreenViewController.createScreenAndUpdateView(screen);
            $("#create_screen_dialog").closeModalForm();
        }
    }

    function screenSelectChanged() {
        ScreenView.updateView(global.screens[ScreenView.getSelectedScreenId()]);
    }

    function screenSelectClicked() {
        ScreenViewController.storeCurrentScreen();
    }

    ScreenViewController.init = function() {
        $("#creatScreenBtn").unbind().click(showCreateScreenDialog);
        if ($("#creatScreenBtnOnPanel").length > 0) {
            $("#creatScreenBtnOnPanel").unbind().click(showCreateScreenDialog);
        }
        $("#screen_select").unbind().change(screenSelectChanged);

        $("#screen_select").click(screenSelectClicked);

    };

    ScreenViewController.createScreenAndUpdateView = function(screen) {
        if (ScreenView.getSelectedScreenId() != 0) {
            ScreenViewController.storeCurrentScreen();
        }

        $("#iphoneBtn_container").empty();

        ScreenViewController.createScreen(screen);
        ScreenView.setLastOptionSelected();
        screenSelectChanged();
    };

    ScreenViewController.createScreen = function(screen) {
        ScreenView.addScreenSelect(screen);
        global.screens[screen.id] = screen;

    };

    ScreenViewController.getCurrentScreen = function() {
        return global.screens[ScreenView.getSelectedScreenId()];
    };



    ScreenViewController.storeCurrentScreen = function() {
        //save previous screen
        if (ScreenView.getSelectedScreenId() != 0) {
            var preScreen = DownloadController.parseCurrentScreen(global.screens[ScreenView.getSelectedScreenId()]);
            global.screens[preScreen.id] = preScreen;
        }
    };




    return ScreenViewController;
} ();

File: branches/project/iPhone_Console_2_0_0_Alphas/Classes/ScreenViewController.m
/* OpenRemote, the Home of the Digital Home.
 * Copyright 2008-2009, OpenRemote Inc.
 * 
 * See the contributors.txt file in the distribution for a
 * full listing of individual contributors.
 * 
 * This is free software; you can redistribute it and/or modify it
 * under the terms of the GNU General Public License as
 * published by the Free Software Foundation; either version 3.0 of
 * the License, or (at your option) any later version.
 * 
 * This software is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
 * 
 * You should have received a copy of the GNU General Public
 * License along with this software; if not, write to the Free
 * Software Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA
 * 02110-1301 USA, or see the FSF site: http://www.fsf.org.
 */


#import "ScreenViewController.h"
#import "ScreenView.h"
#import "ViewHelper.h"
#import "Definition.h"
#import "NotificationConstant.h"
#import "ServerDefinition.h"
#import "CredentialUtil.h"
#import "ControllerException.h"
#import "DataBaseService.h"

@interface ScreenViewController (Private)

- (void)sendCommandRequest:(int)componentId;
- (void)doNavigate:(Navigate *)navi;

@end



@implementation ScreenViewController

@synthesize screen, polling;


- (void)setScreen:(Screen *)s {
	[s retain];
	[screen release];
	screen = s;
	if ([[screen pollingComponentsIds] count] > 0 ) {
		polling = [[PollingHelper alloc] initWithComponentIds:[[[screen pollingComponentsIds] componentsJoinedByString:@","] retain]];
	}
	
}

- (void)performGesture:(Gesture *)gesture {
	Gesture * g = [screen getGestureIdByGestureSwipeType:gesture.swipeType];
	if (g) {
		if (g.hasControlCommand) {
			[self sendCommandRequest:g.componentId];
		} else if (g.navigate) {
			[self doNavigate:g.navigate];
		}
	}

}

// Implement loadView to create a view hierarchy programmatically.
- (void)loadView {
	ScreenView *view = [[ScreenView alloc] init];

	//set Screen in ScreenView
	[view setScreen:screen];
	[self setView:view];
	[view release];
}

- (void)startPolling {
	[polling requestCurrentStatusAndStartPolling];
}
- (void)stopPolling {
	[polling cancelPolling];
}

- (void)sendCommandRequest:(int)componentId {
	
	if ([[Definition sharedDefinition] password] == nil) {
		[[NSNotificationCenter defaultCenter] postNotificationName:NotificationPopulateCredentialView object:nil];
		return;
	}
	
	
	NSString *location = [[NSString alloc] initWithFormat:[ServerDefinition securedControlRESTUrl]];
	NSURL *url = [[NSURL alloc]initWithString:[location stringByAppendingFormat:@"/%d/%@",componentId,@"swipe"]];
	NSLog([location stringByAppendingFormat:@"/%d/%@",componentId,@"swipe"]);
	
	
	//assemble put request 
	NSMutableURLRequest *request = [[NSMutableURLRequest alloc] init];
	[request setURL:url];
	[request setHTTPMethod:@"POST"];
	
	[CredentialUtil addCredentialToNSMutableURLRequest:request];
	
	URLConnectionHelper *connection = [[URLConnectionHelper alloc]initWithRequest:request  delegate:self];
	
	[location release];
	[url	 release];
	[request release];
	[connection autorelease];	
}

- (void)handleServerErrorWithStatusCode:(int) statusCode {
	if (statusCode != 200) {
		if (statusCode == 401) {
			[Definition sharedDefinition].password = nil;
		}
		
		[ViewHelper showAlertViewWithTitle:@"Send Request Error" Message:[ControllerException exceptionMessageOfCode:statusCode]];	
	}
}

#pragma mark delegate method of NSURLConnection
- (void) definitionURLConnectionDidFailWithError:(NSError *)error {
}


- (void)definitionURLConnectionDidFinishLoading:(NSData *)data {
}

- (void)definitionURLConnectionDidReceiveResponse:(NSURLResponse *)response {
	NSHTTPURLResponse *httpResp = (NSHTTPURLResponse *)response;
	
	[self handleServerErrorWithStatusCode:[httpResp statusCode]];
}


- (void)doNavigate:(Navigate *)navi {
	[[NSNotificationCenter defaultCenter] postNotificationName:NotificationNavigateTo object:navi];
}

- (void)dealoc {
	[polling release];
	//[screen release];
	
	[super dealloc];
}

@end
